kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaRequestHandler.java

```diff
@@ -1540,18 +1543,110 @@ protected void handleFetchRequest(KafkaHeaderAndRequest fetch,
+                new ConcurrentHashMap<>();
+
+        AtomicInteger unfinishedAuthorizationCount = new AtomicInteger(request.fetchData().size());
+        Runnable completeOne = () -> {
+            if (unfinishedAuthorizationCount.decrementAndGet() == 0) {
+                TransactionCoordinator transactionCoordinator = null;
+                if (request.isolationLevel().equals(IsolationLevel.READ_COMMITTED)
+                        && kafkaConfig.isKafkaTransactionCoordinatorEnabled()) {
+                    transactionCoordinator = getTransactionCoordinator();
+                }
+                String namespacePrefix = currentNamespacePrefix();
+                int fetchMaxBytes = request.maxBytes();
+                int fetchMinBytes = Math.min(request.minBytes(), fetchMaxBytes);
+                if (interesting.isEmpty()) {
+                    if (log.isDebugEnabled()) {
+                        log.info("Fetch interesting is empty.");
+                    }
+                    resultFuture.complete(new FetchResponse<>(
+                            Errors.NONE,
+                            new LinkedHashMap<>(erroneous),
+                            ((Integer) THROTTLE_TIME_MS.defaultValue),
+                            request.metadata().sessionId()));
+                } else {
+                    MessageFetchContext context = MessageFetchContext
+                            .get(this, transactionCoordinator, maxReadEntriesNum, namespacePrefix,
+                                    getKafkaTopicManagerSharedState(), this.executor, fetch);
+                    this.getReplicaManager().fetchMessage(
+                            request.maxWait(),
+                            fetchMinBytes,
+                            fetchMaxBytes,
+                            interesting,
+                            request.isolationLevel(),
+                            context
+                    ).thenAccept(resultMap -> {
+                        LinkedHashMap<TopicPartition, FetchResponse.PartitionData<Records>> partitions =
+                                new LinkedHashMap<>();
+                        resultMap.forEach((tp, data) -> {
+                            partitions.put(tp, data.toPartitionData());
+                        });
+                        partitions.putAll(erroneous);
+                        resultFuture.complete(new ResponseCallbackWrapper(new FetchResponse<>(
+                                Errors.NONE,
+                                partitions,
+                                ((Integer) THROTTLE_TIME_MS.defaultValue),
+                                request.metadata().sessionId()), () ->
+                                resultMap.forEach((__, readRecordsResult) -> {
+                                    if (readRecordsResult.decodeResult() != null) {
+                                        readRecordsResult.decodeResult().recycle();
+                                    }
```

Comment:
Could it be `null`? If there is a race condition, I think there's also a chance that `readRecordsResult.decodeResult()` is not null in line 1608 but it became `readRecordsResult.decodeResult()` is null in line 1609.

Reply:
The `readRecordsResult#decodeResult ` can be null because we will get an "Errors.NONE" error if there are no more entries to read.